File: 0016.3-sum-closet/16.3-sum-closest.go
package main

import (
	"sort"
)

/*
 * @lc app=leetcode id=16 lang=golang
 *
 * [15] 3Sum Closest
 */
func threeSumClosest(nums []int, target int) int {
	sort.Ints(nums)
	n := len(nums)
	closest := nums[0] + nums[1] + nums[2]
	for i := 0; i <= n-3; i++ {
		l := i + 1
		r := n - 1
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			if abs(sum-target) < abs(closest-target) {
				closest = sum
			}
			if sum >= target {
				r--
			} else {
				l++
			}
		}
	}
	return closest
}

func abs(target int) int {
	if target < 0 {
		return -target
	}
	return target
}

File: 0016.3-sum-closet/16.3-sum-closest.py
#
# @lc app=leetcode id=16 lang=python3
#
# [16] 3Sum Closest
#

# @lc code=start


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        nums.sort()
        closest = nums[0]+nums[1]+nums[2]
        for i in range(n-2):
            l, r = i + 1, n-1
            while l < r:
                s = nums[i]+nums[l]+nums[r]
                if abs(s-target) < abs(closest-target):
                    closest = s
                if s > target:
                    r -= 1
                else:
                    l += 1
        return closest
# @lc code=end
